x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/action/TransportMlInfoAction.java

```diff
@@ -20,26 +21,45 @@
 import org.elasticsearch.xpack.core.ml.datafeed.DatafeedConfig;
 import org.elasticsearch.xpack.core.ml.job.config.AnalysisLimits;
 import org.elasticsearch.xpack.core.ml.job.config.Job;
+import org.elasticsearch.xpack.ml.process.NativeController;
+import org.elasticsearch.xpack.ml.process.NativeControllerHolder;
 
+import java.io.IOException;
 import java.util.HashMap;
 import java.util.Map;
+import java.util.concurrent.TimeoutException;
 import java.util.function.Supplier;
 
 public class TransportMlInfoAction extends HandledTransportAction<MlInfoAction.Request, MlInfoAction.Response> {
 
     private final ClusterService clusterService;
+    private final Map<String, Object> nativeCodeInfo;
 
     @Inject
-    public TransportMlInfoAction(TransportService transportService, ActionFilters actionFilters, ClusterService clusterService) {
+    public TransportMlInfoAction(TransportService transportService, ActionFilters actionFilters,
+                                 ClusterService clusterService, Environment env) {
         super(MlInfoAction.NAME, transportService, actionFilters, (Supplier<MlInfoAction.Request>) MlInfoAction.Request::new);
         this.clusterService = clusterService;
+
+        try {
+            NativeController nativeController = NativeControllerHolder.getNativeController(clusterService.getNodeName(), env);
+            assert nativeController != null;
```

Comment:
This is complicated because in internal cluster tests there is no native controller.

To get the internal cluster tests to work I think you'll have to do something like:

```
    if (nativeController != null) {
        nativeCodeInfo = nativeController.getNativeCodeInfo();
    } else {
        nativeCodeInfo = Collections.emptyMap();
    }
```

or if you want to be more rigorous:

```
    if (MachineLearningField.AUTODETECT_PROCESS.get(env.settings())) {
        NativeController nativeController = NativeControllerHolder.getNativeController(clusterService.getNodeName(), env);
        assert nativeController != null;
        nativeCodeInfo = nativeController.getNativeCodeInfo();
    } else {
        nativeCodeInfo = Collections.emptyMap();
    }
```

It makes me realise that moving this out of the feature set and into a normal transport action makes it possible to just have the native controller as a standard plugin component instead of having that horrible and complicated singleton holder.  I'll refactor that in a followup.

Reply:
I went with your first suggestion for now (see f740ccf), and added a comment to remember to do the conversion from the holder once we've removed native code from xpack info.